File: app/src/main/java/com/mmdev/meetups/ui/custom/ProgressButton.java
package com.mmdev.meetups.ui.custom;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.LinearInterpolator;
import android.view.animation.RotateAnimation;
import android.view.animation.ScaleAnimation;
import android.view.animation.Transformation;

import com.mmdev.meetups.R;

/**
 * Created by lumingmin on 6/3/16.
 */
public class ProgressButton extends View {


	private Paint paintRectF;
	private Paint paintText;

	private Paint paintPro;

	private int mPadding = 0;

	private float mSpac = 0;

	private ProgressButtonAnim mProgressButtonAnim;
	private ScaleAnimation mProgressScaleAnim;
	private RotateAnimation mProgressRotateAnim;
	private RectF mRectF = new RectF();
	private RectF mRectFPro = new RectF();

	private int progerssButtonDuration = 200;
	public void setProgressButtonAnim(int duration)
	{ this.progerssButtonDuration = duration;}

	private int scaleAnimationDuration = 300;
	public void setScaleAnimationDuration (int duration)
	{ this.scaleAnimationDuration = duration;}

	private int bgColor;
	public void setBgColor(int color)
	{ this.bgColor = color; }

	private int textColor;
	public void setTextColor(int color)
	{ this.textColor = color; }

	private int proColor;
	public void setProColor(int color)
	{ this.proColor = color; }

	private String text;
	public void setButtonText(String s) {
		this.text = s;
		invalidate();
	}

	private boolean mStop = false;

	public ProgressButton(Context context, AttributeSet attrs) {
		super(context, attrs);
		TypedArray a = context.getTheme().obtainStyledAttributes(
				attrs,
				R.styleable.ProgressButton,
				0, 0);
		try {
			bgColor = a.getColor(R.styleable.ProgressButton_backgroundColor, Color.BLUE);
			proColor = a.getColor(R.styleable.ProgressButton_progressColor, Color.WHITE);
			text = a.getString(R.styleable.ProgressButton_text);
			textColor = a.getColor(R.styleable.ProgressButton_textColor, Color.WHITE);

		} finally {
			a.recycle();
		}
		initPaint();
	}

	private void initPaint() {
		int mStrokeWidth = dip2px(2);
		mPadding = dip2px(2);
		mProgressButtonAnim = new ProgressButtonAnim();
		mProgressRotateAnim = new RotateAnimation(0f, 360f, Animation.RELATIVE_TO_SELF,
				0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
		mProgressRotateAnim.setRepeatCount(-1);
		mProgressRotateAnim.setInterpolator(new LinearInterpolator());
		mProgressRotateAnim.setFillAfter(true);
		paintRectF = new Paint();
		paintRectF.setAntiAlias(true);
		paintRectF.setStyle(Paint.Style.FILL);
		paintRectF.setStrokeWidth(mStrokeWidth);

		paintText = new Paint();
		paintText.setAntiAlias(true);
		paintText.setStyle(Paint.Style.FILL);
		paintText.setTextSize(dip2px(15));
		paintPro = new Paint();
		paintPro.setAntiAlias(true);
		paintPro.setStyle(Paint.Style.STROKE);
		paintPro.setStrokeWidth((float) mStrokeWidth / 2);

	}

	@Override
	protected void onDraw(Canvas canvas) {
		super.onDraw(canvas);
		paintText.setColor(textColor);
		paintRectF.setColor(bgColor);
		paintPro.setColor(proColor);


		mRectF.left = mPadding + mSpac;
		mRectF.top = mPadding;
		mRectF.right = getMeasuredWidth() - mPadding - mSpac;
		mRectF.bottom = getMeasuredHeight() - mPadding;
		float mRadius = (float) (getMeasuredHeight() - 2 * mPadding) / 2;

		canvas.drawRoundRect(mRectF, mRadius, mRadius, paintRectF);

		if (mRectF.width() == mRectF.height() && !mStop) {
			setClickable(true);

			mRectFPro.left = getMeasuredWidth() / 2.0f - mRectF.width() / 4;
			mRectFPro.top = getMeasuredHeight() / 2.0f - mRectF.width() / 4;
			mRectFPro.right = getMeasuredWidth() / 2.0f + mRectF.width() / 4;
			mRectFPro.bottom = getMeasuredHeight() / 2.0f + mRectF.width() / 4;
			float startAngle = 0f;
			canvas.drawArc(mRectFPro, startAngle, 100, false, paintPro);
		}

		if (mSpac < (getMeasuredWidth() - getMeasuredHeight()) / 2.0f)
			canvas.drawText(text,
					getMeasuredWidth() / 2.0f - getFontlength(paintText, text) / 2.0f,
					getMeasuredHeight() / 2.0f + getFontHeight(paintText, text) / 3.0f,
					paintText);

	}

	public void startAnim() {
		mStop = false;
		setClickable(false);
		if (mProgressButtonAnim != null) {
			clearAnimation();
			mProgressButtonAnim.setDuration(progerssButtonDuration);
		}
		startAnimation(mProgressButtonAnim);
	}

	public void startProAnim() {
		if (mProgressRotateAnim != null) {
			clearAnimation();
			int rotateAnimationDuration = 400;
			mProgressRotateAnim.setDuration(rotateAnimationDuration);
		}
		startAnimation(mProgressRotateAnim);
	}

	public void stopAnim(final OnStopAnim mOnStopAnim) {
		clearAnimation();
		mStop = true;
		invalidate();
		if (mProgressScaleAnim != null)
			clearAnimation();
		else {
			DisplayMetrics displayMetrics = new DisplayMetrics();
			getDisplay().getMetrics(displayMetrics);
			int width = displayMetrics.widthPixels;

			mProgressScaleAnim = new ScaleAnimation(1.0f, (float)width / getMeasuredHeight() * 3.5f,
					1.0f, (float)width / getMeasuredHeight() * 3.5f,
					Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
		}

		mProgressScaleAnim.setDuration(scaleAnimationDuration);
		startAnimation(mProgressScaleAnim);
		mProgressScaleAnim.setAnimationListener(new Animation.AnimationListener() {
			@Override
			public void onAnimationStart(Animation animation) { }

			@Override
			public void onAnimationEnd(Animation animation) {
				mOnStopAnim.Stop();
				clearAnimation();
				mSpac = 0;
				invalidate();
			}

			@Override
			public void onAnimationRepeat(Animation animation) { }
		});
	}


	private class ProgressButtonAnim
			extends Animation {
		@Override
		protected void applyTransformation(float interpolatedTime, Transformation t) {
			super.applyTransformation(interpolatedTime, t);
			mSpac = (getMeasuredWidth() - getMeasuredHeight()) / 2.0f * interpolatedTime;
			invalidate();
			if (interpolatedTime == 1.0f) startProAnim();
		}
	}

	public int dip2px(float dpValue) {
		final float scale = getContext().getResources().getDisplayMetrics().density;
		return (int) (dpValue * scale + 0.5f);
	}

	public float getFontlength(Paint paint, String str) {
		Rect rect = new Rect();
		paint.getTextBounds(str, 0, str.length(), rect);
		return rect.width();
	}

	public float getFontHeight(Paint paint, String str) {
		Rect rect = new Rect();
		paint.getTextBounds(str, 0, str.length(), rect);
		return rect.height();

	}

	public interface OnStopAnim {
		void Stop();
	}


}

File: app/src/main/java/com/mmdev/roove/ui/custom/ProgressButton.kt
/*
 * Created by Andrii Kovalchuk
 * Copyright (c) 2019. All rights reserved.
 * Last modified 04.12.19 19:13
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.mmdev.roove.ui.custom

import android.content.Context
import android.content.res.TypedArray
import android.graphics.*
import android.graphics.Paint.Style
import android.util.AttributeSet
import android.util.DisplayMetrics
import android.view.View
import android.view.animation.*
import android.view.animation.Animation.AnimationListener
import com.mmdev.roove.R.styleable

class ProgressButton: View {
	private var paintRectF: Paint? = null
	private var paintText: Paint? = null
	private var paintPro: Paint? = null
	private var mPadding = 0
	private var mSpac = 0f
	private var mProgressButtonAnim: ProgressButtonAnim? = null
	private var mProgressScaleAnim: ScaleAnimation? = null
	private var mProgressRotateAnim: RotateAnimation? = null
	private val mRectF = RectF()
	private val mRectFPro = RectF()

	private var bgColor = 0
	fun setBgColor(color: Int) {
		bgColor = color
	}

	private var textColor = 0
	fun setTextColor(color: Int) {
		textColor = color
	}

	private var progressColor = 0
	fun setProgressColor(color: Int) {
		progressColor = color
	}

	private var buttonText: String? = null
	fun setButtonText(s: String?) {
		buttonText = s
		invalidate()
	}

	private var progressButtonDuration = 200
	fun setProgressButtonDuration(time: Int) {
		progressButtonDuration = time
	}

	private var progressAnimationSpeed = 400
	fun setProgressAnimationSpeed(time: Int) {
		progressAnimationSpeed = time
	}

	private var scaleAnimationDuration = 300
	fun setScaleAnimationDuration(time: Int) {
		scaleAnimationDuration = time
	}

	private var mStarted = false
	private var mStop = false

	constructor(context: Context?): super(context)

	constructor(context: Context, attrs: AttributeSet?): super(context, attrs) {
		val a: TypedArray =
			context.theme.obtainStyledAttributes(attrs, styleable.ProgressButton, 0, 0)
		try {
			bgColor = a.getColor(styleable.ProgressButton_backgroundColor, Color.BLUE)
			buttonText = a.getString(styleable.ProgressButton_text)
			progressColor = a.getColor(styleable.ProgressButton_progressColor, Color.WHITE)
			textColor = a.getColor(styleable.ProgressButton_textColor, Color.WHITE)
		} finally {
			a.recycle()
		}
		initPaint()
	}

	fun startAnim() {
		if (!mStarted) {
			mStarted = true
			mStop = false
			isClickable = false
			if (mProgressButtonAnim != null) {
				clearAnimation()
				mProgressButtonAnim!!.duration = progressButtonDuration.toLong()
			}
			startAnimation(mProgressButtonAnim)
		}
	}

	fun stopAnim(mOnStopAnim: OnStopAnim) {
		clearAnimation()
		mStop = true
		invalidate()
		if (mProgressScaleAnim != null) clearAnimation()
		else {
			val displayMetrics = DisplayMetrics()
			display.getMetrics(displayMetrics)
			val width = displayMetrics.widthPixels
			mProgressScaleAnim = ScaleAnimation(1.0f,
			                                    width.toFloat() / measuredHeight * 3.5f,
			                                    1.0f,
			                                    width.toFloat() / measuredHeight * 3.5f,
			                                    Animation.RELATIVE_TO_SELF,
			                                    0.5f,
			                                    Animation.RELATIVE_TO_SELF,
			                                    0.5f)
		}
		mProgressScaleAnim!!.duration = scaleAnimationDuration.toLong()
		startAnimation(mProgressScaleAnim)
		mProgressScaleAnim!!.setAnimationListener(object: AnimationListener {
			override fun onAnimationStart(animation: Animation?) {
			}

			override fun onAnimationEnd(animation: Animation?) {
				clearAnimation()
				mOnStopAnim.stop()
				mSpac = 0f
				invalidate()
			}

			override fun onAnimationRepeat(animation: Animation?) {
			}
		})
	}

	private fun initPaint() {
		val mStrokeWidth = dip2px(2f)
		mPadding = dip2px(2f)
		mProgressButtonAnim = ProgressButtonAnim()
		mProgressRotateAnim = RotateAnimation(0f,
		                                      360f,
		                                      Animation.RELATIVE_TO_SELF,
		                                      0.5f,
		                                      Animation.RELATIVE_TO_SELF,
		                                      0.5f)
		mProgressRotateAnim!!.repeatCount = -1
		mProgressRotateAnim!!.interpolator = LinearInterpolator()
		mProgressRotateAnim!!.fillAfter = true
		paintRectF = Paint()
		paintRectF!!.isAntiAlias = true
		paintRectF!!.style = Style.FILL
		paintRectF!!.strokeWidth = mStrokeWidth.toFloat()
		paintText = Paint()
		paintText!!.isAntiAlias = true
		paintText!!.style = Style.FILL
		paintText!!.textSize = dip2px(15f).toFloat()
		paintPro = Paint()
		paintPro!!.isAntiAlias = true
		paintPro!!.style = Style.STROKE
		paintPro!!.strokeWidth = mStrokeWidth.toFloat() / 2
	}

	override fun onDraw(canvas: Canvas) {
		super.onDraw(canvas)
		paintText!!.color = textColor
		paintRectF!!.color = bgColor
		paintPro!!.color = progressColor
		mRectF.left = mPadding + mSpac
		mRectF.top = mPadding.toFloat()
		mRectF.right = measuredWidth - mPadding - mSpac
		mRectF.bottom = measuredHeight - mPadding.toFloat()
		val mRadius = (measuredHeight - 2 * mPadding).toFloat() / 2
		canvas.drawRoundRect(mRectF, mRadius, mRadius, paintRectF!!)
		if (mRectF.width() == mRectF.height() && !mStop) {
			isClickable = true
			mRectFPro.left = measuredWidth / 2.0f - mRectF.width() / 4
			mRectFPro.top = measuredHeight / 2.0f - mRectF.width() / 4
			mRectFPro.right = measuredWidth / 2.0f + mRectF.width() / 4
			mRectFPro.bottom = measuredHeight / 2.0f + mRectF.width() / 4
			val startAngle = 0f
			canvas.drawArc(mRectFPro, startAngle, 100f, false, paintPro!!)
		}
		if (mSpac < (measuredWidth - measuredHeight) / 2.0f) canvas.drawText(buttonText!!,
		                                                                     measuredWidth / 2.0f - getFontLength(
				                                                                     paintText,
				                                                                     buttonText) / 2.0f,
		                                                                     measuredHeight / 2.0f + getFontHeight(
				                                                                     paintText,
				                                                                     buttonText) / 3.0f,
		                                                                     paintText!!)
	}

	private fun progressAnim() {
		if (mProgressRotateAnim != null) {
			clearAnimation()
			mProgressRotateAnim!!.duration = progressAnimationSpeed.toLong()
		}
		startAnimation(mProgressRotateAnim)
	}

	private fun dip2px(dpValue: Float): Int {
		val scale = context.resources.displayMetrics.density
		return (dpValue * scale + 0.5f).toInt()
	}

	private fun getFontLength(paint: Paint?, str: String?): Float {
		val rect = Rect()
		paint!!.getTextBounds(str, 0, str!!.length, rect)
		return rect.width().toFloat()
	}

	private fun getFontHeight(paint: Paint?, str: String?): Float {
		val rect = Rect()
		paint!!.getTextBounds(str, 0, str!!.length, rect)
		return rect.height().toFloat()
	}

	private inner class ProgressButtonAnim: Animation() {
		override fun applyTransformation(interpolatedTime: Float, t: Transformation) {
			super.applyTransformation(interpolatedTime, t)
			mSpac = (measuredWidth - measuredHeight) / 2.0f * interpolatedTime
			invalidate()
			if (interpolatedTime == 1.0f) progressAnim()
		}
	}

	interface OnStopAnim {
		fun stop()
	}
}